### DISCUSSION TITLE
Future ideas for next features in v4.x

### DISCUSSION THREAD

**Discussion by ghiscoding** [ghiscoding/slickgrid-universal#1241] [Ideas]
- [x] 1. migrate [Excel-Builder-Webpacker](https://github.com/ghiscoding/excel-builder.js) fork to TypeScript and provide CJS/ESM builds with Tree Shaking (ref PR #1383)
  - [x] move the code to a new repo [Excel-Builder-Vanilla](https://github.com/ghiscoding/excel-builder-vanilla)
  - [x] drop `Q` and `Lodash` dependencies and possibly replace `JSZip` with `fflate` 
- [x] 2. replace `node-archiver` (to create bundle zip used in Salesforce) by either JSZip or `fflate` (ref PR #1387)
   - [x] need to be tested in Salesforce
- [x] 3. [autocompleter](https://github.com/kraaden/autocomplete) is not fully ESM, I think it could be fixed with a simple PR on the project since it already uses Rollup but currently only build as `umd`, we simply need to add Rollup `es` build 
   - [PR](https://github.com/kraaden/autocomplete/pull/112) creates, just waiting for author's review & merge. I might create a temp fork if I don't have any answer a month from now
   - waited 1.5 months and decided to eventually release [autocompleter-es](https://www.npmjs.com/package/autocompleter-es) until my PR is merge/release (which of course got merged 1 day after I created the fork lol)
- [x] 4. Dark Mode theme (ref PR #1407)
- [ ] 5. possible Vue port
- [x] 6. CustomTooltip plugin only works against 1 cell content but doesn't work well if we have a cell with multiple icons that each have tooltips (e.g. Example 11, 22)... (ref PR #1448)
- [x] 7. deprecate editor model in a way to add back original SlickGrid editor type
   - we could rename the SlickGrid `editor` option to `editorFactory` (or `editorClass`) so that wouldn't be a breaking change to users
- [x] 8. we could show ms-select clear button for clearing the filter (see [showClear](https://ghiscoding.github.io/multiple-select-vanilla/#/options24) option) 
   - only show as a `filterOptions` to Example 11-12
- [x] 9. small Dark Mode tweaks
  - [x] add Flatpickr `box-shadow: 0 2px 6px rgb(255 255 255 / 20%)` to make the picker pop a little more
  - [x] see if we can use SVG for ms-select to make it a bit more modern (like Bulma/Bootstrap Dropdown)
  - [x] tweak uncheck column color
  - [x] we barely see the checkbox border after new version in [Example 12](https://ghiscoding.github.io/slickgrid-universal/#/example12)
  - [x]  small styling issue in Example 7 Dark Mode when filtering date/input, the filter text color is gray instead of white. Should also change SF button primary color to more lighten color (longText/compositeEditor), use same color as header bar border bottom color
- [x] 10. review again package `exports`, use same as ms-select lib, see [are the types wrong](https://arethetypeswrong.github.io/?p=multiple-select-vanilla%403.1.0)
- [x] 11. might be nice to have a way to add default settings for external editor/filter (autocomplete, ms-select, flatpickr, slider)
   - ref PR #1470
   - perhaps something along the lines of (we would have 2 props `defaultEditorOptions` and `defaultFilterOptions`)
```ts
this.gridOptions = {
  defaultEditorOptions: { 
    autocompleter: { debounceWaitMs: 150 }, 
    date: { minDate: 'today' },
    longText: { cols: 50, rows: 5 } 
  }
}
```

### New Discussion
Breaking Change or Postponed Changes were moved to this other Discussion [Roadmap to 5.0](https://github.com/ghiscoding/slickgrid-universal/discussions/1482)

**Comment by ghiscoding** [comment]
hello @zewa666 I'd like some feedback on number ~12~ (now 7) from your side. I'd like to rename the SlickGrid `editor`, not the Slickgrid-Universal one but really the one SlickGrid uses. What's the difference you may ask? The difference is that in Slickgrid-Universal (or upstream repos), the `editor` is where we can add extra properties like `minValue`, `collection`, ... However for SlickGrid the `editor` is really the Editor Class, so that would be `Editors.LongText` class for example (so in Slickgrid-Universal, it's the `editor.model` property). The fact that both projects used the same prop name caused some conflicts which I had to address by internally by copying the user's editor into my own internal ref `internalColumnEditor` and overwriting the `editor` with `editor.model` so that SlickGrid work as intended. Basically I went that way to make it easier for the users because I didn't think of a better prop name `editor` to expose to the users (and also because I used the same for `filter` prop as well).

https://github.com/ghiscoding/slickgrid-universal/blob/21c76cc9d921ad34516bd38070afd791ff55de56/packages/vanilla-bundle/src/components/slick-vanilla-grid-bundle.ts#L1455-L1479

However, now that Slickgrid-Universal is a standalone, I'm thinking about removing this internal properties swap by simply renaming the SlickGrid `editor` to another name, that is where I'd like to have feedback. At first I was thinking about `editorFactory` but in reality it's not really a factory since it's the Editor Class, so maybe a better name is `editorClass`!? Do you have a better name? Then internally, I will copy the `editor.model` into that new name, so possibly doing this: `editor.name => editorClass`

At the end of it, that will cleanup the code and more importantly, that will avoid the need for that internal swap that I was doing (better perf and less error prone). There will be a small change to the users who were referencing `internalColumnEditor` (which equal to the `editor.model`), they will simply be able to use `editor` directly in the future without it being overwritten (which was confusing for some), so it would equal to the same as what they define in their column definitions (which is not the case currently)

so... any name suggestion other than `editorClass` for the one used by SlickGrid? Note that I don't want to be too far from the original SlickGrid because I keep track of whatever PRs is being created and replicate them here when it makes sense

**Reply by zewa666** [reply]
hmm but wouldnt that deviate from the original Slickgrid in quite an important area? I mean not that users are actually using them but there are lots of docs and SO questions referencing that, so renaming it might lead to confusions. or if one wpuld like to port an exisitng slickgridplugin it would be something to keep in mind

aside from that editClass is fine but I'd make it even more explicit by saying slickgridEditorClass

**Reply by ghiscoding** [reply]
yeah that is a side effect, removing confusion of `internalColumnEditor` by adding confusion on why it no longer align with the original SlickGrid. oh well can't win them all. However I still think it's something to do. The other alternative would be to use a different name in my libs but that would be a major breaking change for everyone and a lot of refactoring.

As for the name, I'd like to have both of them starting with the word editor, that is to make it clear that they're not the same but that they work hand in hand. Also the reason of why I'm thinking about this change is because of the new PR #1428 which was opened in SlickGrid by another user and I cannot fully align with his code because of these small `editor` differences.

**Reply by zewa666** [reply]
ouch ok I see. yeah changing it in slickgriduniversal would be a huge breaking change. so I guess your proposal is the one to go with for now

**Reply by ghiscoding** [reply]
I don't care much about the code refactoring inside the lib, but I do care of the code refactoring to the end user. It's already super slow for users to adopt newer version, I imagine that adding a huge code refactoring on their side would make them wait longer to transition to newer version. So the least impact seems to be what I suggest above and I could go ahead right away with this approach and just keep a copy of the `internalColumnEditor` (with a `@deprecated` note) that could be removed in next major. So if I wish to proceed then the user call either `internalColumnEditor` or `editor`, they would both return the same ref (which is not the case today since accessing the `editor` after the grid is created will currently give you the SlickGrid editor class, which is very confusing for many users and I'm trying to solve this)

For example
https://github.com/ghiscoding/slickgrid-universal/blob/21c76cc9d921ad34516bd38070afd791ff55de56/examples/vite-demo-vanilla-bundle/src/examples/example07.ts#L610-L622

Take a look at https://snyk.io/advisor/npm-package/angular-slickgrid, it's still unbelievable to see that the most popular version is still the version before Slickgrid-Universal came on. Why?? Probably because of... "if it ain't broken, don't fix it" approach, even if it doesn't make sense since the Angular version they use are EOL as well... oh well

![image](https://github.com/ghiscoding/slickgrid-universal/assets/643976/d38deb72-2034-469d-9cb5-cd2f5329adde)

**Reply by zewa666** [reply]
welcome to enterprise insanity 😅

sounds like a good battle plan.
I'll take a look at the custom tooltips for multi Buttons once I'm done with my current workload. maybe you could create a separate issue and describe how you'd envision it

**Reply by ghiscoding** [reply]
haha yeah it is what it is, we can't fight it lol

For the tooltip, it's a little tricky for a couple of reasons. I created the plugin with the idea of 1 tooltip to find (most often via a `title` attribute) and I also position the tooltip as per the cell position instead of the target element. For the future, I think that it should position itself via the targeted element but I'm not sure how that's going to work, even the arrow is created via CSS and if it doesn't align with the target then it might look weird. Also the fact that I design the plugin for a single tooltip expectation, this might require some code refactoring to support multi-tooltip on the same cell.

**Reply by ghiscoding** [reply]
Tooltip plugin now support multiple targets on the same cell as in [Example 22](https://ghiscoding.github.io/slickgrid-universal/#/example22), so I think I'm done with everything that required fixing before the next major version. 

For the next major, I really wish that I could replace `moment-mini` with something else but my hope is not at the highest on that one though. I'll give a try to Luxon but if the build output becomes larger then our current build size then I might not go with it, I want to be fully ESM compatible as much as possible but not at all cost. I also would rather wait for the next Angular release (probably around May 20th) to align with them

**Reply by zewa666** [reply]
I cant really speak for others but I personally wouldnt care about the bigger size (70kb right?) since I personally used slickgrid only in either corporate or backend admin frontends. But I get your point. waiting for the next ng major might be a good idea so we could perhaps also update the samples for the new template syntax while at it.

all in all awesome work, and tomorrow my large project starts, so I'm super thrilled to get my hands on all the new stuff.

**Reply by ghiscoding** [reply]
hmm I'm not following much of the new Angular stuff, if you want to modify some of the examples go ahead but I don't think that I will myself ;)

I just tried a quick date parsing last night with Luxon and so far not really impressed, even though some of its author came from Moment, they don't even use the same parsing token (e.g. YYYY-MM-DD is yyyy-MM-dd in Luxon) and it seem to require a lot of different method to deal with parsing and formatting (similar to date-fns I guess). I might give another try at DayJS, they are closer to Moment than any other lib but it's barely maintained now and I remember Zulu wasn't working in strict mode, though that might not be a big issue. I wish date-fns would fix that PR I pointed out... man at the end of the day all I need is parsing/formatting with specific tokens provided to force a certain format for both parsing/formatting, that's all I need in a datagrid (I don't care about any of the date locale stuff)

**Reply by zewa666** [reply]
to be honest I never understood the convention of casing tokens. according to the ISO definition it all upper caps for the date tokens and lower caps for the time part. but its just a recommendation used for samples. the icu (collection of locales) on the other hand uses the later format, as Luxon https://icu4c-demos.unicode.org/icu-bin/locexp?_=en_CA&d_=de&d_l=en same with languages like C#. I get your frustration but ultimately those details are dictated by the lib used, so I'd merely add a note to the token examples of the respective docs

**Reply by ghiscoding** [reply]
yeah I know, it should be totally transparent to the user since these are mostly internal function mapping that I have and the user just defines a `type.dateEuro` and such, then the mapping function assigns the correct token. My frustration is mostly that I still can't find a good replacement for MomentJS. My plan for the long run is to be full ESM eventually (because one day or another ViteJS will be ESM only) and Moment is mostly the last hurdle left to convert (well apart from Flatpickr and DOM Purify but they are less of a problem)

**Reply by zewa666** [reply]
so if you'd fork moment-mini and add a simple esm wrapper on top of cjs would that work instead?

**Reply by ghiscoding** [reply]
ah yes I don't know why I never thought of that but yes that would work, perhaps this [@esm-bundle/moment](https://www.npmjs.com/package/@esm-bundle/moment?activeTab=versions) package. I also read their blog post [An ESM bundle for any NPM package](https://medium.com/@joeldenning/an-esm-bundle-for-any-npm-package-5f850db0e04d) about `@esm-bundle`, it's interesting, so I might give that a try, that might be the easiest solution if it works :)

**Comment by ghiscoding** [comment]
on another subject, I've done some research about Temporal and it doesn't (and never will) have any parser so even if it's released tomorrow, I don't think I could use it out of the box. At some point I tried migrating to DayJS but it has problem with Zulu and most importantly it doesn't have an ESM builds (Luxon has the same problem). Last weekend, I tried to migrate to `date-fns`, it was going ok until I came across the missing non-strict mode in a comment that I described [here](https://github.com/date-fns/date-fns/issues/1569#issuecomment-1989672073), so it looks like I'm stuck with MomentJS for the short term. I wish that they would implement the `strict` flag but that issue has been opened since 2019, so I'm not sure that it will be implemented any time soon (but let's keep dreaming). The non-strict is a real blocker in a datagrid environment

**Reply by zewa666** [reply]
hmm from a quick look it seems luxon does support esm https://github.com/moment/luxon/blob/895a74916a89f816de7c98e912ae9140c6327897/package.json#L11 and can parse zulu times with parseIso + setZone mixes https://github.com/moment/luxon/issues/988

**Reply by ghiscoding** [reply]
For Luxon, it's not quite true, they apparently have a problem with their build and it's not fully ESM (at least not fully tree shackable) as per this [issue](https://github.com/moment/luxon/issues/854) that was never resolved. So I don't want to go with a migration that will not fully work and/or increase the size (I use `moment-mini` which is not that big at 57Kb minified). At this point, it seems that every libs I tried have their own different problems :(

Nonetheless, Luxon is the only one I didn't try yet, I might as well give it a try and see for myself... to be continued

**Reply by ghiscoding** [reply]
so.... on this MomentJS saga, I found out that Moment already transpiles an ESM build since 2020 as per this [commit](https://github.com/moment/moment/commit/87994b745c20febf378ccd8f2dc190cd8d225020), so I was thinking... let's use that then!? Well it's not exactly the best since the biggest reason I switched to `moment-mini` was because by default and without any tricky WebPack configuration, all Moment locales will be included (and there's a **lot** of them). So I switched to `moment-mini` a while ago for that reason. The problem with `moment-mini` is that they only include the CJS build (`moment.js` from the project root), so... I created yet another repo [`moment-tiny`](https://github.com/ghiscoding/moment-tiny) which is similar to moment-mini but with both CJS and ESM builds... and finally I can remove Moment from Angular's `allowedCommonJsDependencies` and it works. More info on [`moment-tiny`](https://github.com/ghiscoding/moment-tiny) repo. 

Also note that even if Moment includes an ESM build, it could be better, one of their biggest disadvantage is that they export everything as `default` (which is discouraged by TypeScript and requires `allowSyntheticDefaultImports`)... but anyway, that should suffice for the next couple years. Maybe some day Temporal will be released and someone will create a great parser with all the tokens needed for my project (someone can dream lol)... or maybe `date-fns` will address the missing non-strict parsing. 

So anyway, I can finally close the subject for now. I'm wondering if I should postpone that for the next major version though since some user might reference `moment-mini` in their project... **but** I'm really tempted to go with just minor version since I'm not crazy about waiting 2 months and technically speaking the user will have the dependency in their `package.json` if they import a dep, so what do you think? Also version adoption is super slow by users of my libs so.. (ref PR #1456)
